File: 项目/OC项目/百思不得姐(oc版本)/百思不得姐/百思不得姐/Classes/Main(主要)/View/QJLTabBar.m
//
//  QJLTabBar.m
//  百思不得姐
//
//  Created by 金亮齐 on 2016/12/6.
//  Copyright © 2016年 醉看红尘这场梦. All rights reserved.
//

#import "QJLTabBar.h"
#import "QJLPublishView.h"
#import "UIView+QJLExtension.h"

@interface QJLTabBar()

@property (nonatomic,weak) UIButton *publishButton;

@end

@implementation QJLTabBar

-(instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundImage = [UIImage imageNamed:@"tabbar-light"];
        //设置发布按钮
        UIButton *publishButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [publishButton setBackgroundImage:[UIImage imageNamed:@"tabBar_publish_icon"] forState:UIControlStateNormal];
        [publishButton setBackgroundImage:[UIImage imageNamed:@"tabBar_publish_click_icon"] forState:UIControlStateSelected];
        [publishButton addTarget:self action:@selector(publishButtonClick) forControlEvents:UIControlEventTouchUpInside];
        publishButton.size = publishButton.currentBackgroundImage.size;
        [self addSubview:publishButton];
        self.publishButton  = publishButton;
    }
    return self;
}

-(void)publishButtonClick{
    [QJLPublishView show];
}

-(void)layoutSubviews {
    
    [super layoutSubviews];
    //设置中间按钮的位置
    CGFloat width = self.width;
    CGFloat heigth = self.height;
    self.publishButton.center = CGPointMake(width * 0.5, heigth * 0.5);
    
    CGFloat buttonY = 0;
    CGFloat buttonW = width / 5;
    CGFloat buttonH = heigth;
    NSInteger index = 0;
    for (UIView *button in self.subviews){
        //如果是系统的按钮，继续执行
        if (![button isKindOfClass:NSClassFromString(@"UITabBarButton")]) continue;
        CGFloat buttonX = buttonW * ((index > 1) ? (index + 1) : index);
        button.frame = CGRectMake(buttonX, buttonY, buttonW, buttonH);
        //索引+1
        index ++;
    }
}
@end

File: 项目/Swift项目/百思不得姐(Swift版本)/百思不得姐/百思不得姐/Classes/Main/View/QJLTabBar.swift
//
//  QJLTabBar.swift
//  百思不得姐
//
//  Created by 金亮齐 on 2016/12/23.
//  Copyright © 2016年 醉看红尘这场梦. All rights reserved.
//

import UIKit

class QJLTabBar: UITabBar {
    
    let plusButton = UIButton.init(type: .Custom)
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        /*添加加号按钮*/
        plusButton.setBackgroundImage(UIImage(named: "tabBar_publish_icon"), forState: .Normal)
        plusButton.setBackgroundImage(UIImage(named: "tabBar_publish_click_icon"), forState: .Highlighted)
        plusButton.addTarget(self, action: #selector(QJLTabBar.presentView), forControlEvents: .TouchUpInside)
        addSubview(plusButton)
        
        backgroundImage = UIImage.init(named: "tabbar-light")
    }
    
    func presentView() {
        UIApplication.sharedApplication().keyWindow?.rootViewController?.presentViewController(PublishController(), animated: false, completion: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let width = frame.width * 0.2
        var buttonIndex:CGFloat = 0
        
        plusButton.sizeToFit()
        plusButton.center = CGPoint(x: self.center.x,y: self.frame.height * 0.5)
        
        for item in subviews {
            if item.isKindOfClass(NSClassFromString("UITabBarButton")!) {
                item.frame = CGRect(x: width * (buttonIndex > 1 ? buttonIndex + 1 : buttonIndex), y: 0, width: width, height: item.frame.height)
                
                buttonIndex++
            }
        }
    }

}
